### PR TITLE
Bootstrapped performance

### PR DESCRIPTION
**Pull request by Jeanselme**
Add a bootstrapping evaluation for the estimation of confidence intervals

### CODE REVIEW

**Comment by Jeanselme** [comment]
I agree, how would you go about it? I am afraid that rerunning the different models with different seeds or even cross-validation is computationally infeasible with such large models

**Comment by mmcdermott** [comment]
> I agree, how would you go about it? I am afraid that rerunning the different models with different seeds or even cross-validation is computationally infeasible with such large models

Well, the first thing to acknowledge is that re-running models with different seeds or splits should be seen as out of scope for MEDS-Evaluation. MEDS Evaluation should start after predictions have been made, and merely evaluate in an appropriate manner. 

The question for us is what to do around the fact that uncertainty should be quantified without deviating from that API. 

There are a couple of options.

1. Do nothing -- have MEDS-Evaluation only be for pointwise evaluations, and make upstream users produce samples of predictions and labels and run the tool themselves many times to assess whatever uncertainty they want. This is a surprisingly reasonable option given the complexity here, and warrants consideration.
2. Only return (but clearly delineate) bootstrapping derived uncertainty estimates. If we are sufficiently clear that this does not cover everything, then that is still a useful option. My concern with this approach is that it will be hard to be "sufficiently clear" given the inadequate landscape on this point.
3. Make the input API clearly have options to express different kinds of repeatedly sampled runs so that uncertainty can be calculated in various ways and reported appropriately. Then, if a user wants to compute a larger notion of uncertainty, they can do so using the API, and it is there responsibility to determine the computational feasibility.

I'm not sure which is the best option, but those are what stand out to me at first glance. What are your thoughts, @Jeanselme?

**Comment by Jeanselme** [comment]
I was thinking of having the bootstrapping options as in 2 or 3 (open to implement any other relevant approaches) - currently, I propose a different function to allow the user to decide to (i) bootstrap estimate and avoid having to recode it or (ii) obtain a point-based estimate

**Comment by mmcdermott** [comment]
> I was thinking of having the bootstrapping options as in 2 or 3 (open to implement any other relevant approaches) - currently, I propose a different function to allow the user to decide to (i) bootstrap estimate and avoid having to recode it or (ii) obtain a point-based estimate

I think this is fine for now, but we need to be very clear. I also think that we should think more deeply and lay out some issues about how (if at all) we want to account for other aspects of uncertainty here. This may also be relevant, though it is AI generated and I haven't validated it fully yet: https://docs.google.com/document/d/1ZxsfKpiQZ-dYa_STJLUZYiZ8uRyeWJiKeL3LITbyTjw/edit?usp=sharing

File: tests/test_binary_classification.py [hunk 1]
```diff
@@ -24,42 +25,71 @@
...
+
+
+def test_main_bootstrapped_binary_classification():
+    binary_classification_main(SAMPLE_CONFIG)
+    # TODO potentially use temporary files
+    with open(Path(SAMPLE_CONFIG.output_dir) / "results_boot.json") as f:
+        actual_output = json.load(f)
+        print(actual_output)
+
+    assert actual_output == EXPECTED_OUTPUT_BOOTSTRAPPED
```
Comment (kamilest) [inline tests/test_binary_classification.py:91]: I'd use pytest.approx to compare these in case there are precision-related errors in floating point values rather than anything implementation-related

File: tests/test_binary_classification.py [hunk 2]
```diff
@@ -24,42 +25,71 @@
 
 BINARY_CLASSIFICATION_SMALL = pl.DataFrame(
     {
-        "subject_id": [1, 2, 3],
+        "subject_id": np.arange(5).repeat(2),
```
Comment (kamilest) [inline tests/test_binary_classification.py:28]: Why does the same subject_id have two different predictions for the same exact time? 

(This should not be a valid MEDS schema I think, if it currently is, and if it's not implemented yet we should implement a schema validator that checks for this)

**Review by kamilest — COMMENTED** [review]
This approach to bootstrapping recalculates metrics based on the subsamples of _predictions_, which is the only thing available to the MEDS evaluation library. 

This is one way to do it, but it seems a bit strange to me because in the previous stage of the pipeline exactly the same subjects were used to generate the predictions, so even if we're taking a subset of the test subjects to derive uncertainty estimates, it's basically using the same train set and the same prediction function, and measures variability of the test subjects at best.

It would seem more natural to me to do bootstrapping in the earlier stage of the pipeline, where the resampling is done on the training set, and evaluating on the whole test set, and so the uncertainty is evaluating the actual variability in model's performance, rather than the variability in the test set. However, this can't be done within MEDS evaluation as it depends on the model training stage.

My opinion seems to be aligned with the discussion at https://stats.stackexchange.com/questions/225433/bootstrapping-test-set
